### PR TITLE
feat(server): Support webm videos

### LINKED ISSUE #1248: [Feature] WebM support
**Describe the bug**
When uploading a webm video from the app, the app throws an  "Bad Request" error.

**Task List**

- [x] I have read thoroughly the README setup and installation instructions.
- [ ] I have included my `docker-compose` file.
- [ ] I have included my redacted `.env` file.
- [x] I have included information on my machine, and environment.

**To Reproduce**
1. Enable the backup feature of the app
2. Download the example file from this issue attachments
3. Make sure the folder where the file is downloaded is covered by the backup

**Expected behavior**
The image should be processed like any other video, possibly by keeping the original webm format on the server which as a format is now supported by all major browsers (iOS has partial support, webm behavior will need to be checked)

[WebM browser coverage by caniuse.com](https://caniuse.com/webm)

**Screenshots**
Screenshot_20230102-165702

screenshot of the error

[ave.webm](https://user-images.githubusercontent.com/47613715/210254871-e3a9eda1-e644-49dc-975e-50a1455c0eae.webm)

original webm file

**System**
- Phone OS [Android]: `13`
- Server Version: `v1.40.0`
- Mobile App Version: `v1.40.0 build.63`

### PR DESCRIPTION
**Pull request by samip5**
This PR includes:

- Support webm videos to be uploaded.
- Add a test case for webm

Closes #1248

### CODE REVIEW

File: server/apps/microservices/src/processors/asset-uploaded.processor.ts [hunk 1]
```diff
@@ -40,7 +40,9 @@ export class AssetUploadedProcessor {
 
     // Video Conversion
     if (asset.type == AssetType.VIDEO) {
-      await this.videoConversionQueue.add(JobName.MP4_CONVERSION, { asset });
+      if (asset.mimeType == 'video/quicktime') {
```
Comment (zackpollard) [inline server/apps/microservices/src/processors/asset-uploaded.processor.ts:43]: In my opinion we shouldn't be making these kind of decisions here. The processing of only quicktime files right now is something we are looking to address, which will be done by the video conversion queue. The video conversion queue should just receive all video files and it can make the decision on whether the file actually needs converting or not.

File: server/apps/immich/src/api-v1/asset/asset.service.ts [hunk 2]
```diff
@@ -456,7 +456,7 @@ export class AssetService {
 
         await fs.access(videoPath, constants.R_OK | constants.W_OK);
 
-        if (query.isWeb && asset.mimeType == 'video/quicktime') {
+        if (query.isWeb && asset.mimeType == 'video/quicktime' || query.isWeb && asset.mimeType == "video/x-msvideo") {
```
Comment (alextran1502) [inline server/apps/immich/src/api-v1/asset/asset.service.ts:459]: Is this change related to this PR?

**Comment by alextran1502** [comment]
I've run some tests, WebM video can be played on the Web, and Android, but cannot play be played on iOS. You might have to convert WebM to MP4. Let me know if you can troubleshoot this issue

**Comment by alextran1502** [comment]
> Hey, thanks for the contribution. WebM support makes sense to me, did you test out how the application handles exif from these types of files with these changes?

FYI, the sample file information panel

image

File: server/apps/immich/src/api-v1/asset/asset.service.ts (hunk 2 above)
Comment (samip5) [reply server/apps/immich/src/api-v1/asset/asset.service.ts:459]: Yes, because previously all files went to get transcoded to MP4 and this PR changed it but .avi files need the transcode as well due to browser support not existing.

**Comment by samip5** [comment]
> I've run some tests, WebM video can be played on the Web, and Android, but cannot play be played on iOS. You might have to convert WebM to MP4. Let me know if you can troubleshoot this issue

Unfortunately I don't have an iOS device so I'm not able to troubleshoot it on that platform and thus why I haven't tested it at all there. (Probably possible somehow to create a iOS emulator on a Mac though?)

I think transcode might be needed, but the API response doesn't usually use the transcoded file except for QuickTime video, or am I mistaken about that?

File: server/apps/microservices/src/processors/asset-uploaded.processor.ts (hunk 1 above)
Comment (samip5) [reply server/apps/microservices/src/processors/asset-uploaded.processor.ts]: How should it make that decision?

**Comment by samip5** [comment]
> Hey, thanks for the contribution. WebM support makes sense to me, did you test out how the application handles exif from these types of files with these changes?

I did test it and as Alex showed it does properly process the exif data for/from them too.

File: server/apps/microservices/src/processors/asset-uploaded.processor.ts (hunk 1 above)
Comment (samip5) [reply server/apps/microservices/src/processors/asset-uploaded.processor.ts]: Currently the transcoder seems to transcode every file that is not mp4.

Comment (jrasm91) [reply server/apps/microservices/src/processors/asset-uploaded.processor.ts]: I think we're saying we can still choose to not convert videos, but "skipping" should happen in one place and that place should be the mp4 processor implementation itself.

File: server/apps/immich/src/api-v1/asset/asset.service.ts (hunk 2 above)
Comment (jrasm91) [reply server/apps/immich/src/api-v1/asset/asset.service.ts]: I think this is a good catch - there's basically a discrepancy between when we transcode a video and when we use the transcoding. Can you make a new function `isWebPlayable(mimeType: string)`. Maybe we can put it in `asset-upload.config.ts` for now, which returns true for all types besides `video/mp4`? And then use that function here (`asset.service.ts`) as well as in the mp4 processor?

Comment (samip5) [reply server/apps/immich/src/api-v1/asset/asset.service.ts]: I have now moved it to a central location and transcoder uses that too.

File: server/apps/microservices/src/processors/asset-uploaded.processor.ts (hunk 1 above)
Comment (samip5) [reply server/apps/microservices/src/processors/asset-uploaded.processor.ts]: I have removed the change here, but I'm wondering how should we handle iOS as it doesn't support webm direct playback apparently?

**Comment by samip5** [comment]
I have now tested my changes with the iOS simulator, and found the same as Alex, so how do we want to handle it?
It would also be great if the iOS app identified itself when sending a request to the api that it's in fact the iOS client that's requesting it rather than `Web`.

**Comment by samip5** [comment]
I believe that this can be merged, as #1370 will better handle the transcoding aspect.
